src/azure-cli-core/azure/cli/core/telemetry.py

```diff
@@ -0,0 +1,133 @@
+import json
+import os
+import sys
+
+from applicationinsights import TelemetryClient
+from applicationinsights.exceptions import enable
+from azure.cli.core import __version__ as core_version
+
+_DEBUG_TELEMETRY = 'AZURE_CLI_DEBUG_TELEMETRY'
+client = {}
+data = []
+APPLICATION = {}
+ARGCOMPLETE_ENV_NAME = None
+
+def set_application(application, argcomplete_env_name):
+    global APPLICATION #pylint: disable=global-statement
+    APPLICATION = application
+    global ARGCOMPLETE_ENV_NAME #pylint: disable=global-statement
+    ARGCOMPLETE_ENV_NAME = argcomplete_env_name
+
+def init_telemetry():
+    try:
+        instrumentation_key = '02b91c82-6729-4241-befc-e6d02ca4fbba'
+
+        global client #pylint: disable=global-statement
+        client = TelemetryClient(instrumentation_key)
+
+        client.context.application.id = 'Azure CLI'
+        client.context.application.ver = core_version
+        client.context.user.id = hash(getpass.getuser())
+
+        enable(instrumentation_key)
+    except Exception as ex: #pylint: disable=broad-except
+        # Never fail the command because of telemetry, unless debugging
+        if _debugging():
+            raise ex
+
+def user_agrees_to_telemetry():
+    # TODO: agreement, needs to take Y/N from the command line
+    # and needs a "skip" param to not show (for scripts)
+    return True
+
+def log_event(event_name, **kwargs):
+    """
+    IMPORTANT: do not log events with quotes in the name, properties or measurements;
+    those events may fail to upload.  Also, telemetry events must be verified in
+    the backend because successful upload does not guarentee success.
+    """
+    try:
+        event_name = _remove_quotes(event_name)
+        for key, value in kwargs.items():
+            kwargs[key] = _remove_quotes(value)
+
+        source = 'az'
+        if os.environ.get('CONTINUOUS_INTEGRATION') and os.environ.get('TRAVIS'):
+            source = 'CI'
+        elif ARGCOMPLETE_ENV_NAME in os.environ:
+            source = 'completer'
+
+        props = {
+            'time': str(time.time()),
+            'x-ms-client-request-id': APPLICATION.session['headers']['x-ms-client-request-id'],
+            'command': APPLICATION.session.get('command', None),
+            'version': core_version,
+            'source': source
+            }
+
+        if kwargs:
+            props.update(**kwargs)
+
+        data.append({
+            'name': event_name,
+            'properties': props
+            })
+    except Exception as ex: #pylint: disable=broad-except
+        # Never fail the command because of telemetry, unless debugging
+        if _debugging():
+            raise ex
+
+def _remove_quotes(s):
+    if isinstance(s, str):
+        return s.replace("'", '_').replace('"', '_')
+    return s
+
+def _make_safe(s):
+    for c in '$%^&|':
+        s = s.replace(c, '_')
+    return s
+
+def _debugging():
+    return (os.environ.get('CONTINUOUS_INTEGRATION') and os.environ.get('TRAVIS')) \
```

Comment:
This can be extract out to a single method and reuse with `log_event` which has the same line